fix memory leak on JRuby 1.x

jakelandis opened the pull request:
It seems that on JRuby 1.x (tested jruby-1.7.27), the outer loop when coded as Java forEach causes JRuby to leak some internal state. Specifically the NonBlockingHashMapLong$CHM dominates the memory. See the referenced bug for a reproduction case. The commit changes the outer loop to a standard for loop to avoid this bug.

Fixes #135

yaauie commented on lib/logstash/filters/grok/timeout_enforcer.rb:
```diff
@@ -56,9 +56,9 @@ def stop!
   def cancel_timed_out!
     now = java.lang.System.nanoTime # save ourselves some nanotime calls
-    @threads_to_start_time.forEach do |thread, start_time|
+    @threads_to_start_time.each do |entry|
```
Since we're only using the key portion of each `Entry`, should we send `Map#keySet()`, and iterate over that instead? Would doing so have the same effect?

~~~ diff
-    @threads_to_start_time.each do |thread, start_time|
+    @threads_to_start_time.keySet.each do |thread|
~~~

jakelandis commented:
@yaauie - thanks!  

I updated to use the keySet, I agree it reads much better. 

I also updated the commit message (and this PR's first comment) to help clarify. I meant Java 8's lambda based forEach instead of the standard for loop.

jakelandis commented:
I ran a test overnight with this code and the 5.6 branch of code.  The same test that helped to identify this issue,  48 workers with a generator input and multiple groks.  Prior to this change, ~5hours in there was a noticeable memory issue, and after ~12 hours Logstash was so un-responsive YourKit wouldn't even stay attached. 

Now: 
<img width="518" alt="image" src="https://user-images.githubusercontent.com/976291/36487035-9097b442-16e5-11e8-89d4-ea404f178a8f.png">

The big drop is a manual GC, and it came back to the same level it was at the start.

jakelandis commented:
@yaauie - thanks. Odd indeed.  Since this only presents itself on JRuby 1.x (which i believe is EOL) I didn't push too hard to understand what is going on under the covers.
